### PR TITLE
Enhance `FunctionExpression` type inference

### PR DESCRIPTION
**Pull request by HerrCai0907**
can infer following `FunctionExpression` type.
```typescript
export function test(arr: i32[]): string[] {
  return arr.map((x: i32) => x.toString());
}
```
- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

File: src/resolver.ts [hunk 1]
```diff
@@ -2637,9 +2637,17 @@ export class Resolver extends DiagnosticEmitter {
     ) {
       // (x) => ret, infer return type accordingt to `ret`
       const expr = (<ExpressionStatement>body).expression;
-      const type = this.resolveExpression(expr, ctxFlow, ctxType, reportMode);
+      let signatureReference = assert(functionType.getSignature());
+      // create a temp flow to resolve expression
+      let tempFlow = Flow.createParent(ctxFlow.actualFunction);
+      assert(signatureReference.parameterTypes.length ==  signature.parameters.length);
+      for (let i = 0, k = signature.parameters.length;i<k;i++) {
```
Comment (MaxGraey) [inline src/resolver.ts:2644]:
```suggestion
      for (let i = 0, k = signature.parameters.length; i < k; i++) {
```

File: src/resolver.ts [hunk 2]
```diff
@@ -2637,9 +2637,17 @@ export class Resolver extends DiagnosticEmitter {
...
+        const parameter = signature.parameters[i];
```
Comment (MaxGraey) [inline src/resolver.ts:2645]:
```suggestion
      let tempFlow = Flow.createParent(ctxFlow.actualFunction);
      let parameters = signature.parameters;
      assert(signatureReference.parameterTypes.length ==  parameters.length);
      for (let i = 0, k = parameters.length; i < k; i++) {
        const parameter = parameters[i];
```

File: src/resolver.ts [hunk 3]
```diff
@@ -2637,9 +2637,18 @@ export class Resolver extends DiagnosticEmitter {
     ) {
       // (x) => ret, infer return type accordingt to `ret`
       const expr = (<ExpressionStatement>body).expression;
-      const type = this.resolveExpression(expr, ctxFlow, ctxType, reportMode);
+      let signatureReference = assert(functionType.getSignature());
+      // create a temp flow to resolve expression
+      let tempFlow = Flow.createParent(ctxFlow.actualFunction);
+      let parameters = signature.parameters;
+      assert(signatureReference.parameterTypes.length ==  parameters.length);
```
Comment (dcodeIO) [inline src/resolver.ts:2644]:
```suggestion
      assert(signatureReference.parameterTypes.length == parameters.length);
```

File: src/resolver.ts [hunk 4]
```diff
@@ -2637,9 +2637,18 @@ export class Resolver extends DiagnosticEmitter {
     ) {
       // (x) => ret, infer return type accordingt to `ret`
       const expr = (<ExpressionStatement>body).expression;
-      const type = this.resolveExpression(expr, ctxFlow, ctxType, reportMode);
+      let signatureReference = assert(functionType.getSignature());
+      // create a temp flow to resolve expression
+      let tempFlow = Flow.createParent(ctxFlow.actualFunction);
+      let parameters = signature.parameters;
+      assert(signatureReference.parameterTypes.length == parameters.length);
```
Comment (dcodeIO) [inline src/resolver.ts:2644]: Can this assert potentially fail? Iirc, function expressions can omit arguments. `Array#forEach` for example is defined as `forEach(fn: (value: T, index: i32, array: Array<T>) => void): void`, and it is sufficient to use it as `.forEach(value => ...)`.

Comment (HerrCai0907) [reply src/resolver.ts:2644]: in `resolveFunctionType`, `functionType.getSignature()` is generated by `parameters`. So the length must be same.
But it is guarantee by the `resolveFunctionType`. So I add a assert here to for quick find the bug when `resolveFunctionType` change its behavior.

Comment (dcodeIO) [reply src/resolver.ts:2644]: Can you add a comment? :)

**Review by MaxGraey — APPROVED** [review]

**Comment by MaxGraey** [comment]
Thanks!

**Pull request merged**